### ISSUE TITLE
[3.x] Crash when reparenting some nodes and deleting former parent.

### ISSUE THREAD

**Issue by m4nu3lf** [godotengine/godot#57097]
### Godot version

30fe807fa9eeca2039b8c25ee7775b23c12ce3b4

### System information

OS: Arch Linux, CPU: Ryzen 5, GUP: RX 480

### Issue description

The issue seems to be caused by reparenting collision shapes and deleting the previous rigid body parent.
When, after reparenting, the collision shape is added to the tree and then deleted/removed from the tree, a crash or error happens.

The error is:
```
E 0:00:00.387   erase: Condition "p_I->data != this" is true. Returned: false
  <C++ Source>  ./core/list.h:150 @ erase()
  <Stack Trace> CollisionShapeReparentingCrash.gd:13 @ _ready()
```

I have done some debugging and it seems related to the owner being the former parent and being deleted. I added some print() statements to the MRP. Godot always crashes when I try to print the owner of the reparented child after it was deleted.

### Steps to reproduce

Open the MRP in the editor and run the scene named `CollisionShapeReparentingCrash.tscn`
A crash will happen at the last print() statement.
Remove the print statement and you'll get an error. I was not able to reproduce it exactly as in my main project but there I get a crash instead.

The crash/error doesn't seem to show up if the rigid body is not a sub-scene or with other types of node.

### Minimal reproduction project

[collision_shape_reparenting_crash.zip](https://github.com/godotengine/godot/files/7921291/collision_shape_reparenting_crash.zip)

**Comment by m4nu3lf** [comment]
@Chaosus I think this should be for the  3.5 milestone. I haven't tested it on 4.0

**Comment by m4nu3lf** [comment]
So, I think I fixed it locally. I added some lines of code before this [line](https://github.com/godotengine/godot/blob/3.x/scene/main/node.cpp#L3002). This is the code I added:
```
	for (List<Node*>::Element *E = data.owned.front(); E; E = E->next()) {
		E->get()->set_owner(nullptr);
	}
	if (data.owner) {
		data.owner->data.owned.erase(data.OW);
	}
```
Should I create a PR?

**Comment by qarmin** [comment]
```
=================================================================
==35270==ERROR: AddressSanitizer: heap-use-after-free on address 0x619000133d90 at pc 0x000012682ae5 bp 0x7ffdcf595730 sp 0x7ffdcf595720
READ of size 8 at 0x619000133d90 thread T0
    #0 0x12682ae4 in Variant::Variant(Object const*) core/variant.cpp:2168
    #1 0x9433ce0 in MethodBind0RC<Node*>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:593
    #2 0x121cd553 in ClassDB::get_property(Object*, StringName const&, Variant&) core/class_db.cpp:1046
    #3 0x1243a61c in Object::get(StringName const&, bool*) const core/object.cpp:499
    #4 0x12875e66 in Variant::get_named(StringName const&, bool*) const core/variant_op.cpp:1699
    #5 0x1eb2916 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:636
    #6 0x1cf3941 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1214
    #7 0x1cf3bf7 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1220
    #8 0xca1e284 in Node::_notification(int) scene/main/node.cpp:146
    #9 0x1b9a52d in Node::_notificationv(int, bool) scene/main/node.h:45
    #10 0x1bc4bd5 in Spatial::_notificationv(int, bool) scene/3d/spatial.h:52
    #11 0x12448382 in Object::notification(int, bool) core/object.cpp:927
    #12 0xca1fd39 in Node::_propagate_ready() scene/main/node.cpp:179
    #13 0xca1f590 in Node::_propagate_ready() scene/main/node.cpp:171
    #14 0xca737ea in Node::_set_tree(SceneTree*) scene/main/node.cpp:2592
    #15 0xcb3b088 in SceneTree::init() scene/main/scene_tree.cpp:473
    #16 0x18ac1a9 in OS_X11::run() platform/x11/os_x11.cpp:3697
    #17 0x18150ae in main platform/x11/godot_x11.cpp:55
    #18 0x7fc8d6a2bfcf in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #19 0x7fc8d6a2c07c in __libc_start_main_impl ../csu/libc-start.c:409
    #20 0x1814ccd in _start (/usr/bin/godots+0x1814ccd)

0x619000133d90 is located 16 bytes inside of 1056-byte region [0x619000133d80,0x6190001341a0)
freed by thread T0 here:
    #0 0x7fc8d79ae40f in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:122
    #1 0x129dba7c in Memory::free_static(void*, bool) core/os/memory.cpp:168
    #2 0x19e7194 in void memdelete<Object>(Object*) core/os/memory.h:117
    #3 0x12447396 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:891
    #4 0x126da088 in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1185
    #5 0x1ec4abf in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1046
    #6 0x1cf3941 in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1214
    #7 0x1cf3bf7 in GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1220
    #8 0xca1e284 in Node::_notification(int) scene/main/node.cpp:146
    #9 0x1b9a52d in Node::_notificationv(int, bool) scene/main/node.h:45
    #10 0x1bc4bd5 in Spatial::_notificationv(int, bool) scene/3d/spatial.h:52
    #11 0x12448382 in Object::notification(int, bool) core/object.cpp:927
    #12 0xca1fd39 in Node::_propagate_ready() scene/main/node.cpp:179
    #13 0xca1f590 in Node::_propagate_ready() scene/main/node.cpp:171
    #14 0xca737ea in Node::_set_tree(SceneTree*) scene/main/node.cpp:2592
    #15 0xcb3b088 in SceneTree::init() scene/main/scene_tree.cpp:473
    #16 0x18ac1a9 in OS_X11::run() platform/x11/os_x11.cpp:3697
    #17 0x18150ae in main platform/x11/godot_x11.cpp:55
    #18 0x7fc8d6a2bfcf in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

previously allocated by thread T0 here:
    #0 0x7fc8d79ae808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x129daa39 in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:75
    #2 0x129da94a in operator new(unsigned long, char const*) core/os/memory.cpp:40
    #3 0xc980908 in Object* ClassDB::creator<RigidBody>() (/usr/bin/godots+0xc980908)
    #4 0x121c0377 in ClassDB::instance(StringName const&) core/class_db.cpp:520
    #5 0xf9dd411 in SceneState::instance(SceneState::GenEditState) const scene/resources/packed_scene.cpp:182
    #6 0xfa10c9e in PackedScene::instance(PackedScene::GenEditState) const scene/resources/packed_scene.cpp:1590
    #7 0xf9dc70c in SceneState::instance(SceneState::GenEditState) const scene/resources/packed_scene.cpp:163
    #8 0xfa10c9e in PackedScene::instance(PackedScene::GenEditState) const scene/resources/packed_scene.cpp:1590
    #9 0x19d397b in Main::start() main/main.cpp:2080
    #10 0x1814fde in main platform/x11/godot_x11.cpp:54
    #11 0x7fc8d6a2bfcf in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58

SUMMARY: AddressSanitizer: heap-use-after-free core/variant.cpp:2168 in Variant::Variant(Object const*)

```

**Comment by RandomShaper** [comment]
I'll take a look ASAP.

**Comment by akien-mga** [comment]
Fixed by #59063.
